### PR TITLE
factory context: refactored factory context

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: factory context: refactored factory context
Additional Description:

1. removed `FactoryContextBase`
2. `FactoryContext` won't inherit from `CommonFactoryContext`.

What should I say? Cheers? 😄 

Risk Level: mid.
Testing: n/a.
Docs Changes: n/a. 
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
🍻 finally...

File: source/common/local_reply/local_reply.cc [hunk 1]
```diff
@@ -23,7 +23,7 @@ class BodyFormatter {
         content_type_(Http::Headers::get().ContentTypeValues.Text) {}
 
   BodyFormatter(const envoy::config::core::v3::SubstitutionFormatString& config,
-                Server::GenericFactoryContextImpl context)
+                Server::Configuration::GenericFactoryContext& context)
```
Comment (alyssawilk) [inline source/common/local_reply/local_reply.cc:26]: optional: think it's worth doing a typedef for this one just to avoid the churn here and for folks downstream?

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc [hunk 2]
```diff
@@ -17,7 +17,7 @@ namespace Common {
 namespace DynamicForwardProxy {
 
 absl::StatusOr<std::shared_ptr<DnsCacheImpl>> DnsCacheImpl::createDnsCacheImpl(
-    const Server::Configuration::GenericFactoryContext& context,
+    Server::Configuration::GenericFactoryContext& context,
```
Comment (alyssawilk) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:20]: why the change to non-const?

**Review by alyssawilk — APPROVED** [review]
this change is _AMAZING_

I'd bump risk to high just because it's a pretty big refactor.  Feel free to merge as is and typedef in a follow-up someone else can review as I'm sure merge conflicts will be a lot.
